### PR TITLE
Cache token introspection response from OIDC provider

### PR DESCRIPTION
**Pull request by H-Shay**
This PR adds an expiring cache to the result of the `_introspect_token` of the  `MSC3861DelegatedAuth` class. The cache expires after five minutes, and the token introspection response is checked for expiration both when the response is pulled from the cache and when the response is received over network (although I am not sure if the second check is necessary, would the authorizing server ever send an introspection response with an expired `exp` field?).

A follow-on PR will add an endpoint for receiving revocations from the authorizing server.

### CODE REVIEW

File: synapse/api/auth/msc3861_delegated.py [hunk 1]
```diff
@@ -144,50 +153,75 @@ async def _introspect_token(self, token: str) -> IntrospectionToken:
...
-            raise ValueError(
-                "The introspection endpoint returned an invalid JSON response."
-            )
+            if not isinstance(resp, dict):
+                raise ValueError(
+                    "The introspection endpoint returned an invalid JSON response."
+                )
+
+            expiration = resp.get("exp", None)
+            if expiration:
```
Comment (H-Shay) [inline synapse/api/auth/msc3861_delegated.py:215]: I am unsure if this is necessary - would the authorizing server ever send an introspection response with an expired `exp` field?

File: synapse/api/auth/msc3861_delegated.py [hunk 2]
```diff
@@ -144,50 +153,75 @@ async def _introspect_token(self, token: str) -> IntrospectionToken:
...
+        # check the cache before doing a request
+        introspection_token = self._token_cache.get(token, None)
+
+        expired = False
+        if introspection_token:
+            # check the expiration field of the token (if it exists)
+            exp = introspection_token.get("exp", None)
+            if exp:
+                time_now = self._clock.time_msec()
+                expired = time_now > exp
```
Comment (erikjohnston) [inline synapse/api/auth/msc3861_delegated.py:165]: To double check, is `exp` in milliseconds or seconds?

File: synapse/api/auth/msc3861_delegated.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/api/auth/msc3861_delegated.py:215]: I think its a worthwhile sanity check. It certainly could happen if there is *significant* clock drift between the two servers, but yeah it shouldn't happen in a healthy deployment

File: synapse/api/auth/msc3861_delegated.py [hunk 3]
```diff
@@ -144,50 +153,75 @@ async def _introspect_token(self, token: str) -> IntrospectionToken:
...
+
+        if not introspection_token or expired:
```
Comment (erikjohnston) [inline synapse/api/auth/msc3861_delegated.py:167]: Rather than indenting can we just return early?

**Review by erikjohnston — COMMENTED** [review]
I think this looks good, thanks!

**Comment by erikjohnston** [comment]
Have requested review from @sandhose just to triple check this is the correct logic.

File: synapse/api/auth/msc3861_delegated.py (hunk 2 above)
Comment (H-Shay) [reply synapse/api/auth/msc3861_delegated.py:165]: Good catch it is actually in seconds: https://datatracker.ietf.org/doc/html/rfc7662#section-2.2

**Comment by H-Shay** [comment]
@erikjohnston per the meeting - are you fine with me going ahead and merging this?

**Comment by erikjohnston** [comment]
Why?

**Comment by erikjohnston** [comment]
By why, I mean yes! Sorry, got distracted writing that one word response....

**Comment by erikjohnston** [comment]
Merge it!

**Comment by H-Shay** [comment]
Merged #16117 into develop, thanks for the review!

**Pull request merged**